cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+    private final SchedulableProgramType programType;
+    private final AtomicBoolean active;
+    private StreamSizeSchedule streamSizeSchedule;
+
+    // Size, in bytes, given by the polling info which serves as a base when comparing polling info
+    private long basePollSize;
+    // Time, in milliseconds, when the previous attribute was computed
+    private long basePollTs;
+
+    // Size, in bytes, of the stream at the last recorded execution
+    private long lastRunSize;
+    // Logical time, in milliseconds, of the last recorded execution
+    private long lastRunTs;
+
+    private StreamSizeScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                   StreamSizeSchedule streamSizeSchedule) {
+      this.programId = programId;
+      this.programType = programType;
+      this.streamSizeSchedule = streamSizeSchedule;
+      this.active = new AtomicBoolean(false);
+    }
+
+    /**
+     * Start a stream size schedule task.
+     *
+     * @param basePollSize base size of the stream to start counting from. This info comes from polling the stream
+     * @param basePollTs time at which the {@code basePollSize} was obtained
+     * @param active {@code true} if the schedule is active, {@code false} if it is suspended
+     * @param persist {@code true} if this information should be persisted, {@code false} otherwise
+     */
+    public void startSchedule(long basePollSize, long basePollTs, boolean active, boolean persist) {
+      LOG.debug("Starting schedule {} with basePollSize {}, basePollTs {}, active {}. Should be persisted: {}",
+                streamSizeSchedule.getName(), basePollSize, basePollTs, active, persist);
+      this.basePollSize = basePollSize;
+      this.basePollTs = basePollTs;
+      this.lastRunSize = -1;
+      this.lastRunTs = -1;
+      this.active.set(active);
+    }
+
+    public boolean isActive() {
+      return active.get();
+    }
+
+    /**
+     * Received stream size information coming from polling.
+     *
+     * @param pollingInfo {@link StreamSizeNotification} info that came from polling the stream using metrics
+     */
+    public void receivedPollingInformation(@Nonnull StreamSizeNotification pollingInfo) {
+      Preconditions.checkNotNull(pollingInfo);
+      if (!active.get()) {
+        return;
+      }
+
+      StreamSizeSchedule currentSchedule;
+      ImmutableMap.Builder<String, String> argsBuilder = ImmutableMap.builder();
+      synchronized (this) {
+        if (pollingInfo.getSize() - basePollSize < toBytes(streamSizeSchedule.getDataTriggerMB())) {
+          return;
+        }
+
+        currentSchedule = streamSizeSchedule;
+        basePollSize = pollingInfo.getSize();
+        basePollTs = pollingInfo.getTimestamp();
+
+        argsBuilder.put(ProgramOptionConstants.SCHEDULE_NAME, streamSizeSchedule.getName());
+        argsBuilder.put(ProgramOptionConstants.LOGICAL_START_TIME, Long.toString(pollingInfo.getTimestamp()));
```

Comment:
Do you realize that a few lines above, you set basePollSize and basePollTs to whatever is in the pollingInfo? So effectively in the arguments, they are always the same.